### PR TITLE
Development bc mu 10210 no lep producers

### PR DESCRIPTION
**Pull request by cgalloni**

### CODE REVIEW

File: Ntuplizer/plugins/JpsiMuNtuplizer.cc [hunk 1]
```diff
@@ -317,400 +318,478 @@ particle_cand JpsiMuNtuplizer::calculateIPvariables(
...
-  std::vector<pat::Muon> muoncollection;
-  std::vector<int> muoncollection_id;
-  muoncollection.clear();
-  muoncollection_id.clear();
-
-  for(size_t imuon = 0; imuon < muons_->size(); ++ imuon){
+    // Second trigger if fist one didn't fire
+    if (!isTriggered){
+        for (unsigned int i = 0, n = HLTtriggers_->size(); i < n; ++i) {
+            if(trigNames.triggerName(i).find("HLT_DoubleMu4_JpsiTrk_Displaced_v")!= std::string::npos){
```
Comment (gitytakahas) [inline Ntuplizer/plugins/JpsiMuNtuplizer.cc]: This seems to be duplicated to L342. You wanted to put another HLT path name here ?

File: Ntuplizer/plugins/JpsiMuNtuplizer.cc [hunk 2]
```diff
@@ -317,400 +318,478 @@ particle_cand JpsiMuNtuplizer::calculateIPvariables(
...
+           // right now we're just checking that there are at least two muons passing our kinematic cuts
+           // so we don't care if there are more than two muons that pass right now
+           if(mu2passpteta) continue;
+           if(muon.pt() < 4) continue;
+           if(TMath::Abs(muon.eta()) > 2.4) continue;
+           if(!(muon.track().isNonnull())) continue;
+           // If we're on the Second muon, then the second muon is what passed 
+           if (mu1passpteta && !mu2passpteta) { mu2passpteta = true;}
+           // If we're on the First muon, then the first muon is what passed
+           if (!mu1passpteta) {mu1passpteta = true;}
```
Comment (gitytakahas) [inline Ntuplizer/plugins/JpsiMuNtuplizer.cc:412]: I don't understand what do you want to do here ...

File: Ntuplizer/plugins/JpsiMuNtuplizer.cc [hunk 3]
```diff
@@ -317,400 +318,478 @@ particle_cand JpsiMuNtuplizer::calculateIPvariables(
...
       
-      if(trigger_dR < 0.1) trigMatch = true;
+            if(trigger_dR < 0.1) trigMatch = true;
+        }
+
+        if(!trigMatch) continue;
+        // If we're on the Second muon, then the second muon is what passed 
+        if (mu1passtrigmatch && !mu2passtrigmatch) { mu2passtrigmatch = true;}
+        // If we're on the First muon, then the first muon is what passed
+        if (!mu1passtrigmatch) {mu1passtrigmatch = true;}
```
Comment (gitytakahas) [inline Ntuplizer/plugins/JpsiMuNtuplizer.cc:481]: Here as well ... I don't understand what do you want to do here

File: Ntuplizer/plugins/JpsiMuNtuplizer.cc [hunk 4]
```diff
@@ -317,400 +318,478 @@ particle_cand JpsiMuNtuplizer::calculateIPvariables(
...
+           if(TMath::Abs(muon.eta()) > 2.4) continue;
+           if(!(muon.track().isNonnull())) continue;
+           // If we're on the Second muon, then the second muon is what passed 
+           if (mu1passpteta && !mu2passpteta) { mu2passpteta = true;}
+           // If we're on the First muon, then the first muon is what passed
+           if (!mu1passpteta) {mu1passpteta = true;}
+           }
+
+       if (mu1passpteta) {nBranches_->cutflow_perevt->Fill(1);}
+       if (mu2passpteta) {nBranches_->cutflow_perevt->Fill(2);}
```
Comment (gitytakahas) [inline Ntuplizer/plugins/JpsiMuNtuplizer.cc:416]: I guess you forget to add, 
if(doCutFlow) nBranches_ -> ...

File: Ntuplizer/plugins/JpsiMuNtuplizer.cc (hunk 1 above)
Comment (cgalloni) [reply Ntuplizer/plugins/JpsiMuNtuplizer.cc]: Hi Yuta, this is the second trigger that was suggested by Palla and Co. My idea was to save events that passed this second trigger, in case the other one is not fired.

File: Ntuplizer/plugins/JpsiMuNtuplizer.cc (hunk 2 above)
Comment (cgalloni) [reply Ntuplizer/plugins/JpsiMuNtuplizer.cc:412]: I'll let Wren comment this one, since she implementted it. But I guess she was just checking the fact that there are 2 muons in the sample passing the kinemtic selection.

File: Ntuplizer/plugins/JpsiMuNtuplizer.cc (hunk 4 above)
Comment (cgalloni) [reply Ntuplizer/plugins/JpsiMuNtuplizer.cc:416]: there is a "doCutFlow_) few lines before.

File: Ntuplizer/plugins/JpsiMuNtuplizer.cc (hunk 1 above)
Comment (gitytakahas) [reply Ntuplizer/plugins/JpsiMuNtuplizer.cc]: Yeah, but the name seems to be exactly the same to me ...: HLT_DoubleMu4_JpsiTrk_Displaced_v*

File: Ntuplizer/plugins/JpsiMuNtuplizer.cc (hunk 2 above)
Comment (gitytakahas) [reply Ntuplizer/plugins/JpsiMuNtuplizer.cc:412]: Hmmm ... but you have almost duplicate of the codes between L397 to L488, which I don't like. Probably, my question is, rather, why do you need cutflow for each of the muon, rather than just J/psi selection ? Why do we need such granularity?

Comment (cgalloni) [reply Ntuplizer/plugins/JpsiMuNtuplizer.cc:412]: @yuta, this was just to check the basic kinematic selection efficiency on the samples that I produced to check the validity of the samples and find out if the gen filters cuts were appropriately placed. In further iterations of the code we could remove this bit and have just the J/psi selection.

Comment (gitytakahas) [reply Ntuplizer/plugins/JpsiMuNtuplizer.cc:412]: ok got it.

File: Ntuplizer/plugins/GenParticlesNtuplizer.cc [hunk 5]
```diff
@@ -18,86 +19,166 @@ GenParticlesNtuplizer::~GenParticlesNtuplizer( void )
...
-  //chunk to remove those events with no jspi if that analysis is chosen
-  std::vector<int> doJpsi_;
-  if(isJpsiEle_) {
-    doJpsi_ = nBranches_->IsJpsiEle;
-  }else if(isJpsiMu_){
-    doJpsi_ = nBranches_->IsJpsiMu;
-  }
+  
 
-  if(size(doJpsi_)==0) return;
```
Comment (gitytakahas) [inline Ntuplizer/plugins/GenParticlesNtuplizer.cc:29]: @cgalloni sorry I missed this before, but why did you remove this line? I thought it would be helpful, not to save all Gen info ... otherwise the file becomes too big ...
